### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by bindi-nagda**
I wanted to draft this PR so that @abarret can look at one of the tests that can't be run due to a bug. See [tests/convective_op/convective_op_2d.uniThn.amrGrid.input](https://github.com/abarret/multiphase-stokes/compare/main...add_tests#diff-9b95e1377ea21b7636bb9248561697ac5f9c223063047ea3e536165a31f2c7cf). This file results in an error we observed before: 

`P=00000:Program abort called in file ``/home/bnagda2015/sfw/samrai/2.4.4/linux-g++-opt/include/GridGeometry.C'' at line 760
P=00000:ERROR MESSAGE: P=00000:Error in hier::GridGeometry<DIM> object with name = CartesianGeometry: in computeMaxGhostWidth():  Cannot add variables and increase maximum ghost width after creating the GridGeometry!
`

This was fixed in the hierarchy integrator class for tests with time stepping whereby we create the convective operators before the patch hierarchy is finished being created so we can specify the max ghost cell widths. I suppose we need to do a similar thing, but inside the convective manager class?

### CODE REVIEW

File: tests/convective_op/convective_op.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        N_un_fcn.setDataOnPatchHierarchy(e_un_idx, e_un_var, patch_hierarchy, 0.0);
+        N_us_fcn.setDataOnPatchHierarchy(e_us_idx, e_us_var, patch_hierarchy, 0.0);
+
+        // Apply convective operator
+        INSVCTwoFluidConvectiveManager convec_op("convec_op",
+                                                 patch_hierarchy,
+                                                 input_db->getDatabase("ConvecOp"),
+                                                 un_bc_coefs,
+                                                 us_bc_coefs,
+                                                 thn_bc_coef.get());
```
Comment (abarret) [inline tests/convective_op/convective_op.cpp:236]: The issue is here. The patch index with the largest ghost cell width needs to be created before the patch hierarchy is created. This line just needs to be moved before line 178

Comment (bindi-nagda) [reply tests/convective_op/convective_op.cpp:236]: I tried that last night, but still ran into the same error. Let me check again though...

Comment (bindi-nagda) [reply tests/convective_op/convective_op.cpp:236]: Results in a seg fault for both uniform and amr grids:

```
Thread 1 "convective_op_2" received signal SIGSEGV, Segmentation fault.
0x00005555555c749a in SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchLevel<2> >::Pointer(SAMRAI::tbox::PointerBase const&) ()
```

Comment (abarret) [reply tests/convective_op/convective_op.cpp:236]: I really hate samrai sometimes. They make it so hard to do RAII.

The issue is some of the members of `INSVCTwoFluidConvectiveManager` rely on the patch hierarchy being fully functional. In particular, `d_hier_sc_data_ops` and `d_hier_cc_data_ops` should be replaced with `std::unique_ptr<>` and be created as needed, probably in `allocateData()`.

Comment (abarret) [reply tests/convective_op/convective_op.cpp:236]: Can you do this? I can also push a change to this branch if you want me to.

Comment (bindi-nagda) [reply tests/convective_op/convective_op.cpp:236]: Yeah, let me try this

**Comment by bindi-nagda** [comment]
The bug from above has now been fixed by using unique pointers to the hierarchy data operations. I'm not certain why this fixed the problem though as I'm new to using the unique_ptr class.

**Comment by abarret** [comment]
`unique_ptr` is a smart pointer that automatically deletes itself when the `unique_ptr` goes out of scope, similar to `SAMRAI::tbox::Pointer` and `std::shared_ptr`. The main difference is that `unique_ptr` can not be copied, meaning only one instantiation of `unique_ptr` can be created. This is in contrast to `shared_ptr` and `Pointer`, where you can have multiple copies.

**Comment by abarret** [comment]
@bindi-nagda Do you want to fix the conflicts or should I?

**Comment by bindi-nagda** [comment]
> @bindi-nagda Do you want to fix the conflicts or should I?

I can take care of those

**Comment by bindi-nagda** [comment]
@abarret This is ready for review.

File: tests/physical_bc/apply_2d.uniThn.amrGrid.output [hunk 2]
```diff
@@ -0,0 +1,17 @@
+|e|_oo = 8.12101
+|e|_2  = 3.07985
+|e|_1  = 2.32166
+
+Error in RHS_un :
+  L1-norm:  1.728856669
+  L2-norm:  2.921730398
+  max-norm: 8.121012094
```
Comment (abarret) [inline tests/physical_bc/apply_2d.uniThn.amrGrid.output:8]: I recall discussing a bug in some implementation, but I don't remember where. Did we fix that?

**Comment by abarret** [comment]
These tests fail on my laptop. I will investigate.

File: tests/physical_bc/apply_2d.uniThn.amrGrid.output (hunk 2 above)
Comment (bindi-nagda) [reply tests/physical_bc/apply_2d.uniThn.amrGrid.output:8]: None of the programs in physical bc exhibit 2nd order convergence. I did not include any tests to check for 2nd order convergence though.

**Comment by bindi-nagda** [comment]
I ran the linter and pushed the changes. Maybe the tests are failing because the formatting of the .input and .output files in `tests/` needs to be adjusted using dos2unix(?) I didn't run into this issue today but it has happened in the past.

**Comment by abarret** [comment]
I'm getting error norms that are about half the size of the output.

**Comment by abarret** [comment]
This is specifically for the following tests
```
    physical_bc/solve_2d.homogeneous.uniGrid.input
    physical_bc/channel_2d.uniThn.uniGrid.input
    physical_bc/solve_2d.inhomogeneous.uniGrid.input
    physical_bc/channel_2d.uniThn.amrGrid.input
    physical_bc/solve_2d.homogeneous.amrGrid.input
    physical_bc/solve_2d.inhomogeneous.amrGrid.input
```

**Comment by bindi-nagda** [comment]
@abarret  I updated the outputs so those should pass for you now. I had to do a git pull on IBAMR master because my last build was from Feb 11th.

**Review by abarret — APPROVED** [review]

File: tests/physical_bc/apply_2d.uniThn.amrGrid.output (hunk 2 above)
Comment (abarret) [reply tests/physical_bc/apply_2d.uniThn.amrGrid.output:8]: So this is by design. The `MultiphaseStaggeredStokesOperator` uses linear extrapolation to fill ghost cell values. A 5 point Laplacian does not converge at the boundary if you use linear extrapolation. If you use quadratic extrapolation, then the operator converges.

**Pull request merged**